src/spxmod/regmod_builder.py

```diff
@@ -0,0 +1,246 @@
+        if self.params[0].linear_upriors:
+            self.linear_umat = asmatrix(
+                sp.csc_matrix(param.linear_upriors[0].mat)
+            )
+
+        # parse constraints
+        cmat = asmatrix(
+            sp.vstack(
+                [sp.identity(self.mat[0].shape[1]), self.linear_umat],
+                format="csr",
+            )
+        )
+        cvec = np.hstack([self.uvec, self.linear_uvec])
+
+        if cmat.size > 0:
+            scale = abs(cmat).max(axis=1).toarray().ravel()
+            valid = ~np.isclose(scale, 0.0)
+            cmat, cvec, scale = cmat[valid], cvec[:, valid], scale[valid]
+            if scale.size > 0:
+                cmat = cmat.scale_rows(1.0 / scale)
+                cvec = cvec / scale
+
+                neg_valid = ~np.isneginf(cvec[0])
+                pos_valid = ~np.isposinf(cvec[1])
+                cmat = sp.vstack(
+                    [-cmat[neg_valid], cmat[pos_valid]], format="csr"
+                )
+                cvec = np.hstack([-cvec[0][neg_valid], cvec[1][pos_valid]])
+
+        self.cmat = asmatrix(sp.csc_matrix(cmat))
+        self.cvec = cvec
+
+    def detach_df(self) -> None:
+        self.data.detach_df()
+        del self.mat
+        del self.uvec
+        del self.gvec
+        del self.linear_uvec
+        del self.linear_gvec
+        del self.linear_umat
+        del self.linear_gmat
+        del self.cmat
+        del self.cvec
+
+    def hessian_from_gprior(self) -> Matrix:
+        """Hessian matrix from the Gaussian prior.
+
+        Returns
+        -------
+        Matrix
+            Hessian matrix.
+
+        """
+        hess = sp.diags(1.0 / self.gvec[1] ** 2, format="csr")
+        if self.linear_gvec.size > 0:
+            hess += (
+                self.linear_gmat.T.scale_cols(1.0 / self.linear_gvec[1] ** 2)
+            ).dot(self.linear_gmat)
+        return asmatrix(hess)
+
+    def fit(
+        self,
+        data: DataFrame,
+        encode: Callable,
+        optimizer: Callable = msca_optimize,
+        **optimizer_options,
+    ) -> None:
+        self.attach_df(data, encode)
+        super().fit(optimizer=optimizer, **optimizer_options)
+        self.detach_df()
+
+    def predict(
+        self,
+        data: DataFrame,
+        encode: Callable,
+        return_ui: bool = False,
+        alpha: float = 0.05,
+    ) -> NDArray:
+        mat = sp.csc_matrix(encode(data))
+        param, coef = self.params[0], self.opt_coefs
+
+        offset = np.zeros(len(data))
+        if param.offset is not None:
+            offset = data[param.offset].to_numpy()
+
+        lin_param = offset + mat.dot(coef)
+        pred = param.inv_link.fun(lin_param)
+
+        if return_ui:
+            if alpha < 0 or alpha > 0.5:
+                raise ValueError("`alpha` has to be between 0 and 0.5")
+            # TODO: explore the sparsity of the variance-covariance matrix
+            vcov = get_vcov(self.opt_hessian, self.opt_jacobian2)
```

Comment:
I think we should either raise or return nans if the model is sparse here, as we aren't computing uncertainty yet in this case.

Reply:
Thanks @AHsu98! Added a cap for number of variables when compute uncertainty. Otherwise will use the dense matrix to do so.